### PR TITLE
Core/Object: Range check

### LINKED ISSUE #23062: CheckRange for gameobject targets
**Description:**

CheckRange needs to include m_targets.GetGOTarget() checks. Allows tons of exploits.

**Current behaviour:**

Doesn't

**Steps to reproduce the problem:**

Use a Key at a big distance on a locked door.

**Branch(es):**

3.3.5

**TC rev. hash/commit:** 

951f130f20ef6e023c5d5c220eefb0ed9544ddef

Suggested:

```diff
diff --git a/src/server/game/Spells/Spell.cpp b/src/server/game/Spells/Spell.cpp
index 63af3ae2e..1624da1bf 100755
--- a/src/server/game/Spells/Spell.cpp
+++ b/src/server/game/Spells/Spell.cpp
@@ -6194,6 +6194,16 @@ SpellCastResult Spell::CheckRange(bool strict) const
             (m_spellInfo->FacingCasterFlags & SPELL_FACING_FLAG_INFRONT) && !m_caster->HasInArc(static_cast<float>(M_PI), target))
             return SPELL_FAILED_UNIT_NOT_INFRONT;
     }
+    
+    if (GameObject* go = m_targets.GetGOTarget())
+    {
+        if (m_caster->GetExactDistSq(go) > maxRange)
+            return SPELL_FAILED_OUT_OF_RANGE;
+
+        if (minRange > 0.0f && m_caster->GetExactDistSq(go) < minRange)
+            return SPELL_FAILED_OUT_OF_RANGE;
+    }
 
     if (m_targets.HasDst() && !m_targets.HasTraj())
     {
```

### PR DESCRIPTION
**Pull request by d0d0**
**Changes proposed:**

-  better range checking on gobs targets

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:**

Closes #23062

**Tests performed:** (Does it build, tested in-game, etc.)

Tested in game on known test cases (in issue #23062)

### CODE REVIEW

File: src/server/game/Entities/GameObject/GameObject.h [hunk 1]
```diff
@@ -318,7 +328,7 @@ class TC_GAME_API GameObject : public WorldObject, public GridObject<GameObject>
         GameObjectValue m_goValue;
 
         int64 m_packedRotation;
-        QuaternionData m_worldRotation;
+        G3D::Quat m_localRotation;
```
Comment (Shauren) [inline src/server/game/Entities/GameObject/GameObject.h]: This must not be changed - no G3D header should ever be included in any of core headers (compile time reasons)

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by Killyana** [comment]
Does it also fix https://github.com/TrinityCore/TrinityCore/issues/15453 ?

**Comment by illfated** [comment]
> Does it also fix #15453 ?

Would be pretty sweet if it does.

**Comment by D0d0** [comment]
Well i am out of PC till monday, so i will check it later

Dňa 12. 4. 2019 o 15:05 užívateľ Trond B. Krokli <notifications@github.com> napísal:

**Comment by Killyana** [comment]
The I tested the PR, the issue that allow players to open doors using the key from far away could be reproduced.
About #15453 the behavior is still the same.

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 2]
```diff
@@ -2581,3 +2585,119 @@ std::string GameObject::GetDebugInfo() const
...
+{
+    if (spell || (spell = GetSpellForLock(player)))
+    {
+        float maxRange = spell->GetMaxRange(spell->IsPositive());
+
+        if (GetGoType() == GAMEOBJECT_TYPE_SPELL_FOCUS)
+            return maxRange * maxRange >= GetExactDistSq(player);
+
+        // Not yet decompiled
+        if (GetGoType() == GAMEOBJECT_TYPE_DOOR)
```
Comment (jackpoz) [inline src/server/game/Entities/GameObject/GameObject.cpp]: could you handle this case with a lower range than infinite ?

File: src/server/game/Entities/GameObject/GameObject.h (hunk 1 above)
Comment (D0d0) [reply src/server/game/Entities/GameObject/GameObject.h]: Can I use them in `.cpp` file?

**Comment by D0d0** [comment]
I did small research on issue #15453 (not tested) and maybe we can replace `IsWithinDistInMap` in `Player.cpp` with new function `GameObject::IsAtInteractDistance(Position const& pos, float radius) const`
```
GameObject* Player::GetGameObjectIfCanInteractWith(ObjectGuid const& guid) const
{
    if (GameObject* go = GetMap()->GetGameObject(guid))
    {
        if (go->IsWithinDistInMap(this, go->GetInteractionDistance()))
            return go;
        ...
    }

    return nullptr;
}
```

**Comment by jackpoz** [comment]
could you take a look at the door case too ?

**Comment by D0d0** [comment]
Since we do not know the exact value for doors, I changed them from infinte to 5.0f

**Comment by Killyana** [comment]
The interaction range with doors is perfect, just #15453 remains.

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 3]
```diff
@@ -2581,3 +2600,119 @@ std::string GameObject::GetDebugInfo() const
...
+        for (auto&& playerSpell : player->GetSpellMap())
+            if (SpellInfo const* spell = sSpellMgr->GetSpellInfo(playerSpell.first))
+                for (auto&& effect : spell->Effects)
+                    if (effect.Effect == SPELL_EFFECT_OPEN_LOCK && ((uint32) effect.MiscValue) == lock->Index[i])
+                        if (effect.CalcValue(player) >= int32(lock->Skill[i]))
+                            return spell;
+    }
+
+    return nullptr;
+}
```
Comment (illfated) [inline src/server/game/Entities/GameObject/GameObject.cpp]: Please add NewLine (`CR/LF`) here.

**Comment by jackpoz** [comment]
should we merge this PR the way it is and investigate the large gossip unrelated issue later on ? @D0d0 did you test the code you posted at https://github.com/TrinityCore/TrinityCore/pull/23179#issuecomment-483961419 ?

**Comment by D0d0** [comment]
I did not do any further research about #23179

**Comment by illfated** [comment]
Just out of curiosity: would it be correct to write the PR title as
**Core/Object: range check**
? (I don't know, I just base my suggestion on the PR content)

**Comment by jackpoz** [comment]
let's see what happens

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Comment by xvwyh** [comment]
Oh wow, github decided not to subscribe me to this PR automatically even though I left a comment here. I missed all the action!

> I did small research on issue #15453 (not tested) and maybe we can replace `IsWithinDistInMap` in `Player.cpp` with new function `GameObject::IsAtInteractDistance(Position const& pos, float radius) const`
> 
> ```
> GameObject* Player::GetGameObjectIfCanInteractWith(ObjectGuid const& guid) const
> {
>     if (GameObject* go = GetMap()->GetGameObject(guid))
>     {
>         if (go->IsWithinDistInMap(this, go->GetInteractionDistance()))
>             return go;
>         ...
>     }
> 
>     return nullptr;
> }
> ```

Yes, this will absolutely fix #15453.

If I had known you had this `GameObject::GetInteractionDistance` method, I would've suggested you move the giant switch statement from `GameObject::IsAtInteractDistance(Player const* player, SpellInfo const* spell)` to `GameObject::GetInteractionDistance` and use it in place of the switch. But the use of `IsAtInteractDistance` would still be pretty much mandatory for any range check to gameobjects.

**Comment by jackpoz** [comment]
@D0d0 while you are at it could you implement the suggestions above , https://github.com/TrinityCore/TrinityCore/commit/32e1de39a26628dcb64bc21ad415afb2ad938925#r33298498 and test again that the code works ?

**Comment by D0d0** [comment]
Yes, changes definitely work for larger GoBs, I will open new PR soon also with https://github.com/TrinityCore/TrinityCore/issues/23205 and https://github.com/TrinityCore/TrinityCore/issues/23204

**Comment by jackpoz** [comment]
please keep https://github.com/TrinityCore/TrinityCore/issues/23204 on another PR though, that has little to do with GameObjects

**Pull request merged**